README.md
# LPLAYER - A simple audio player

**Lplayer** is a simple audio player for simply listening music or whatever you want. The first time I thinked about **lplayer** was for listening podcasts. **lplayer** can set the speed. I normally listen podcasts a 1.8x. So this option was very important for me. But not only. This audio player must have to avoid silences to play podcasts.

**lplayer** can play several audio formats,

* MP3
* OGG
* FLAC
* M4a

After I added the option to play several audio format, I thinked that this audio player must include an equalizer, so you can configure the player as you want. Besides, **lplayer** incorporates some equalizer presets, so it's more easy to configure it.

![](/screenshots/lplayer_01.png)

**lplayer** is thinked as a very simple audio player that it doesn't disturb your concentration. Only to play audio and if it is possible in background.

![](/screenshots/lplayer_02.png)

In order to not disturb, **lplayer** is integrated in the sound menu in Linux Mint Cinnamon and Ubuntu. With latest versions of Ubuntu you will need a extension for it. Lplayer uses the audio menu to display information about the track you are listening. From the menu, you can go to the next or the previous track, play, pause and stop the track.

![](/screenshots/lplayer_03.png)


For more information, please visit our website:

[https://www.atareao.es/aplicacion/lplayer/](lplayer)

* [Code](https://github.com/atareao/lplayer)
* [Bugs](https://bugs.launchpad.net/lplayer)
* [Blueprints](https://blueprints.launchpad.net/lplayer)
* [Translations](https://translations.launchpad.net/lplayer)
* [Answers](https://answers.launchpad.net/lplayer)
* [Page project in Launchpad](https://launchpad.net/lplayer)

![](/screenshots/lplayer_04.png)

# Requirements

Required dependencies:

```
    python3,
    python3-gi,
    python3-dbus,
    python3-requests,
    python3-dbus,
    python3-pil,
    gir1.2-gtk-3.0,
    gir1.2-gdkpixbuf-2.0,
    gir1.2-notify-0.7,
    gir1.2-gstreamer-1.0,
    gir1.2-gst-plugins-base-1.0,
    gstreamer1.0-tools,
    gstreamer1.0-plugins-good,
    gstreamer1.0-plugins-ugly,
    gstreamer1.0-plugins-bad,
    gstreamer1.0-libav,
```

# Install

To install lplayer in Ubuntu, Linux Mint and derivatives, run these commands in a terminal,

```
    sudo add-apt-repository ppa:atareao/lplayer
    sudo apt update
    sudo apt install lplayer
```


debian/changelog
lplayer (0.4.1-0extras20.04.2) focal; urgency=medium

  * Added an indicator play
  * Added shortcuts for player
  * Fixed dependencies
  * Update translations

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 18 Apr 2020 21:15:47 +0200

lplayer (0.3.4-0extras16.04.3) xenial; urgency=medium

  * Added french translation 
  * Fixed a bug with info dialog

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 26 Mar 2018 19:01:21 +0200

lplayer (0.3.4-0extras16.04.1) bionic; urgency=medium

  * Added shorcuts

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Fri, 23 Mar 2018 17:33:07 +0100

lplayer (0.3.3-0extras16.04.3) bionic; urgency=medium

  * Track end detection
  * Clear configuration on new version

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 13 Mar 2018 11:58:20 +0100

lplayer (0.3.2-0extras16.04.1) xenial; urgency=medium

  * Play added tracks

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 12 Mar 2018 10:25:02 +0100

lplayer (0.3.1-0extras16.04.0) xenial; urgency=medium

  * Presets de Foobar2000

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 11 Mar 2018 09:46:06 +0100

lplayer (0.3.0-0extras16.04.0) xenial; urgency=medium

  * Now can add folders
  * Can set position from trackview

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 10 Mar 2018 20:43:32 +0100

lplayer (0.2.5-0extras16.04.2) xenial; urgency=medium

  * Fixed a bug with add_tracks();

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 10 Mar 2018 18:59:07 +0100

lplayer (0.2.5-0extras16.04.0ubuntu1) xenial; urgency=medium

  * Updated to play m4a and flac from desktop

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 28 Feb 2018 18:42:20 +0100

lplayer (0.2.4-0extras16.04.0) xenial; urgency=medium

  * Load tracks in background
  * Save equalizer preset

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 28 Feb 2018 09:21:51 +0100

lplayer (0.2.3-0extras16.04.0) xenial; urgency=medium

  * Stop feature added

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 27 Feb 2018 21:27:59 +0100

lplayer (0.2.2-0extras16.04.0) xenial; urgency=medium

  * Support for mp4 media file

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 27 Feb 2018 20:44:20 +0100

lplayer (0.2.1-0extras16.04.0) xenial; urgency=medium

  * Support for flac media file

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 27 Feb 2018 19:43:52 +0100

lplayer (0.2.0-0extras16.04.0) xenial; urgency=medium

  * Added equalizer presets
  * Updated translations
  * Added some tooltips

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 27 Feb 2018 18:08:12 +0100

lplayer (0.1.12-0extras16.04.0) xenial; urgency=medium

  * removing thumbnails

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 27 Feb 2018 06:55:28 +0100

lplayer (0.1.11-0extras16.04.1) xenial; urgency=medium

  * Updated for Ubuntu themes

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 26 Feb 2018 22:12:03 +0100

lplayer (0.1.10-0extras16.04.0) xenial; urgency=medium

  * Updated info dialog
  * Updated translations

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 26 Feb 2018 19:11:56 +0100

lplayer (0.1.9-0extras16.04.0ubuntu1) xenial; urgency=medium

  * reorder

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 26 Feb 2018 18:20:46 +0100

lplayer (0.1.8-0extras16.04.0) xenial; urgency=medium

  * Focus

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 26 Feb 2018 14:10:46 +0100

lplayer (0.1.7-0extras16.04.0) xenial; urgency=medium

  * Move scroll with audio
  * Autoplay

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 26 Feb 2018 13:14:20 +0100

lplayer (0.1.6-0extras16.04.0) xenial; urgency=medium

  * Start with track
  * Launch from terminal
  * Play sound

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 26 Feb 2018 12:26:29 +0100

lplayer (0.1.5-0extras16.04.3) xenial; urgency=medium

  * First pre-release
  * Fixed some dependencies

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 25 Feb 2018 21:34:10 +0100

lplayer (0.1.5-0extras17.04.3) zesty; urgency=medium

  * Added download all option
  * Added WebUpd8 repository to preinst

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 03 Jan 2018 11:36:24 +0100

lplayer (0.1.4-0extras17.04.0ubuntu1) zesty; urgency=medium

  * Added mutagen

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 16 Aug 2017 15:30:56 +0200

lplayer (0.1.3-0extras17.04.0) zesty; urgency=medium

  * Added support for download youtube lists

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 09 Aug 2017 12:55:39 +0200

lplayer (0.1.2-0extras17.04.0) zesty; urgency=medium

  * Added preferences

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 08 Aug 2017 19:18:58 +0200

lplayer (0.1.1-0extras17.04.0) zesty; urgency=medium

  * Improved README
  * Added dependencies

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 08 Aug 2017 12:26:44 +0200

lplayer (0.1.0-0extras17.04.5) zesty; urgency=medium

  * First release

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 08 Aug 2017 11:59:03 +0200


src/lplayer/lplayer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of lplayer
#
# Copyright (c) 2017-2019 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('Gdk', '3.0')
    gi.require_version('Gio', '2.0')
    gi.require_version('GLib', '2.0')
    gi.require_version('GObject', '2.0')
    gi.require_version('GdkPixbuf', '2.0')
    gi.require_version('Notify', '0.7')
except Exception as e:
    print(e)
    exit(1)
from gi.repository import Gtk
from gi.repository import Gio
from gi.repository import GLib
from gi.repository import GdkPixbuf
from gi.repository import Notify
import sys
import webbrowser
from .mainwindow import MainWindow
from . import comun
from .comun import _


class MainApplication(Gtk.Application):
    def __init__(self):
        Gtk.Application.__init__(
            self,
            application_id='es.atareao.lplayer',
            flags=Gio.ApplicationFlags.HANDLES_COMMAND_LINE
        )
        self.license_type = Gtk.License.GPL_3_0
        self.win = None
        self.new_tracks = []

    def do_shutdown(self):
        Gtk.Application.do_shutdown(self)

    def on_quit(self, widget, data):
        self.quit()

    def do_command_line(self, command_line):
        if command_line is not None:
            print(command_line.get_arguments(), self.win)
            if len(command_line.get_arguments()) > 1:
                self.new_tracks.extend(command_line.get_arguments()[1:])
        self.do_activate()
        return 0

    def do_startup(self):
        print('do_startup')
        Gtk.Application.do_startup(self)

        def create_action(name,
                          callback=self.action_clicked,
                          var_type=None,
                          value=None):
            if var_type is None:
                action = Gio.SimpleAction.new(name, None)
            else:
                action = Gio.SimpleAction.new_stateful(
                    name,
                    GLib.VariantType.new(var_type),
                    GLib.Variant(var_type, value)
                )
            action.connect('activate', callback)
            return action

        self.add_action(create_action("quit", callback=lambda *_: self.quit()))

        self.set_accels_for_action('app.add', ['<Ctrl>A'])
        self.set_accels_for_action('app.open', ['<Ctrl>O'])
        self.set_accels_for_action('app.quit', ['<Ctrl>Q'])
        self.set_accels_for_action('app.about', ['<Ctrl>F'])

        self.set_accels_for_action('app.play', ['Up'])
        self.set_accels_for_action('app.pause', ['Down'])
        self.set_accels_for_action('app.previous', ['Left'])
        self.set_accels_for_action('app.next', ['Right'])

        self.add_action(create_action(
            'new',
            callback=self.on_headbar_clicked))
        self.add_action(create_action(
            'open',
            callback=self.on_headbar_clicked))
        self.add_action(create_action(
            'close',
            callback=self.on_headbar_clicked))
        self.add_action(create_action(
            'save',
            callback=self.on_headbar_clicked))
        self.add_action(create_action(
            'save_as',
            callback=self.on_headbar_clicked))
        '''s
        self.add_action(create_action(
            'download_all',
            callback=self.on_download_all_clicked))
        '''
        self.add_action(create_action(
            'set_preferences',
            callback=self.on_preferences_clicked))
        self.add_action(create_action(
            'goto_homepage',
            callback=lambda x, y: webbrowser.open(
                'http://www.atareao.es/aplicacion/lplayer')))
        self.add_action(create_action(
            'goto_code',
            callback=lambda x, y: webbrowser.open(
                'https://github.com/atareao/lplayer')))
        self.add_action(create_action(
            'goto_bug',
            callback=lambda x, y: webbrowser.open(
                'https://github.com/atareao/lplayer/issues')))
        self.add_action(create_action(
            'goto_sugestion',
            callback=lambda x, y: webbrowser.open(
                'https://blueprints.launchpad.net/lplayer')))
        self.add_action(create_action(
            'goto_translation',
            callback=lambda x, y: webbrowser.open(
                'https://translations.launchpad.net/lplayer')))
        self.add_action(create_action(
            'goto_questions',
            callback=lambda x, y: webbrowser.open(
                'https://answers.launchpad.net/lplayer')))
        self.add_action(create_action(
            'goto_twitter',
            callback=lambda x, y: webbrowser.open(
                'https://twitter.com/atareao')))
        self.add_action(create_action(
            'goto_google_plus',
            callback=lambda x, y: webbrowser.open(
                'https://plus.google.com/\
118214486317320563625/posts')))
        self.add_action(create_action(
            'goto_facebook',
            callback=lambda x, y: webbrowser.open(
                'http://www.facebook.com/elatareao')))
        self.add_action(create_action(
            'goto_donate',
            callback=lambda x, y: webbrowser.open(
                'https://www.atareao.es/donar/')))
        self.add_action(create_action(
            'about',
            callback=self.on_about_activate))
        self.add_action(create_action(
            'quit',
            callback=self.on_quit_activate))
        self.add_action(create_action(
            'none',
            callback=self.do_none))
        self.add_action(create_action(
            'play',
            callback=self.on_play_activate))
        self.add_action(create_action(
            'pause',
            callback=self.on_pause_activate))
        self.add_action(create_action(
            'previous',
            callback=self.on_previous_activate))
        self.add_action(create_action(
            'next',
            callback=self.on_next_activate))

        action_toggle = Gio.SimpleAction.new_stateful(
            "toggle", None, GLib.Variant.new_boolean(False))
        action_toggle.connect("change-state", self.toggle_toggled)
        self.add_action(action_toggle)

        lbl_variant = GLib.Variant.new_string("h3")
        new_action = Gio.SimpleAction.new_stateful("new",
                                                   lbl_variant.get_type(),
                                                   lbl_variant)
        new_action.connect("activate", self.activate_radio)
        new_action.connect("change-state", self.toggle_heading)
        self.add_action(new_action)

        action_heading = Gio.SimpleAction.new_stateful(
            "heading",
            GLib.VariantType.new("s"),
            GLib.Variant("s", "h1"))
        action_heading.connect("activate", self.activate_radio)
        action_heading.connect("change-state", self.toggle_heading)
        self.add_action(action_heading)

    def activate_radio(self, widget, action, parameter=None):
        self.win.menu['lists'].set_label(action.get_string())
        widget.set_state(action)

    def heading(self, action):
        print(action)

    def toggle_heading(self, action, state):
        print(action, state)

    def do_activate(self):
        print('do_activate')
        if self.win is None:
            self.win = MainWindow(self)
            self.win.show_all()
        self.add_window(self.win)
        self.win.present()
        if len(self.new_tracks) > 0:
            self.win.add_tracks_in_background(self.new_tracks)
            self.new_tracks = []

    def action_clicked(self, action, variant):
        print(action, variant)
        if variant:
            action.set_state(variant)

    def on_headbar_clicked(self, action, optional):
        self.win.on_toolbar_clicked(action, action.get_name())

    def on_support_clicked(self, widget, optional):
        pass

    def on_preferences_clicked(self, action, optional):
        self.win.on_preferences_clicked(None)

    def on_download_all_clicked(self, action, optional):
        self.win.on_download_all_clicked(None)

    def on_quit_activate(self, widget, optional):
        self.win._sound_menu_quit()

    def on_play_activate(self, widget, optional):
        self.win._sound_menu_play()

    def on_pause_activate(self, widget, optional):
        self.win._sound_menu_pause()

    def on_previous_activate(self, widget, optional):
        self.win._sound_menu_previous()

    def on_next_activate(self, widget, optional):
        self.win._sound_menu_next()


    def on_about_activate(self, widget, optional):
        ad = Gtk.AboutDialog(comun.APPNAME, self.win)
        ad.set_name(comun.APPNAME)
        ad.set_version(comun.VERSION)
        ad.set_copyright('Copyrignt (c) 2018-2019\nLorenzo Carbonell')
        ad.set_comments(_('A minimal audio player for Linux'))
        ad.set_license('''
MIT License

Copyright (c) 2012-2018 Lorenzo Carbonell Cerezo <a.k.a. atareao>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
''')
        ad.set_website('http://www.atareao.es')
        ad.set_website_label('http://www.atareao.es')
        ad.set_authors([
            'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
        ad.set_documenters([
            'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
        ad.set_translator_credits('\
Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>\n')
        ad.set_program_name(comun.APPNAME)
        ad.set_logo(GdkPixbuf.Pixbuf.new_from_file(comun.ICON))
        ad.run()
        ad.destroy()

    def do_none(self, *args):
        pass

    def toggle_toggled(self, action, state):
            action.set_state(state)
            Gtk.Settings.get_default().set_property(
                "gtk-application-prefer-dark-theme", state)


def main(args):
    Notify.init('lplayer')
    app = MainApplication()
    app.run(args)


if __name__ == "__main__":
    main(sys.argv)
